File: src/CPU/Memory.ts
export type Memory = {
  read: (address: number) => number
  write: (address: number, data: number) => void
  load: (arr: Uint32Array) => void
  LENGTH: number
}

/**
 * Create a new Area of memory with a uint32 array, which can read/write data and load programs
 * @returns a Memory Type object with read/write and load capabilities
 */
export function createMemory(length = 2000): Memory {

  const LENGTH = length
  const MEMORY = new Uint32Array(LENGTH)

  // memory read wrapper
  const read = (address: number) => MEMORY[address]
  // memory write wrapper
  const write = (address: number, data: number) => { MEMORY[address] = data }

  const load = (arr: Uint32Array) => MEMORY.set(arr)

  return { read, write, load, LENGTH }
}

File: virtual-machine/src/vm/memory.rs
const mem_length: usize = 2048;
const STACK_SIZE: usize = 2048;

pub struct Memory {
    mem: [i32; mem_length],
}

impl Memory {
    pub fn new() -> Memory {
        Memory {
            mem: [0; mem_length],
        }
    }

    pub fn read(&self, address: usize) -> i32 {
        self.mem[address]
    }

    pub fn write(&mut self, address: usize, value: i32) {
        self.mem[address] = value;
    }
}
